### ISSUE TITLE
Performance and maintenance problems with AST and Schema traversals

### ISSUE THREAD

**Issue by gkesler** [graphql-java/graphql-java#904]
Currently there are many recursive algorithms all over the places in the code written by different developers and with different style. Not mentioning maintenance problems due to the nature of recursive algorithms and they way how they are used, there are also very important practical aspects that should be addressed in the long run:
* it is very hard to implement breadth-first tree traversal of a tree using call stack recursion; the resulting code becomes very difficult to maintain. Example: https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/batched/BatchedExecutionStrategy.java
* performance overhead caused by call stack recursion
* direct dispatch mechanism based on the runtime type check of a **tree node** -> **cast it to the runtime type** -> **execution of the target method** has its performance and maintenance overheads. Performance overhead is caused by heavy use of reflection (*instanceOf* and cast to a target type). To illustrate the later, if a new type is introduced, there is NO compile time support to detect places in the where this new type needs to be handled. This becomes very important for the 3rd parties deeply integrating with this project's API.

**Comment by bbakerman** [comment]
Do you think your new TreeTraversal code could be applied to BatchedExecutionStrategy?

That code is pre me and is a bit of a mind bender in trying to understand?

**Comment by gkesler** [comment]
yes, I would like to refactor all current execution strategies using this mechanism.
BatchTraversalStrategy is the one that I'd like to refactor most of all

**Comment by gkesler** [comment]
plus there are a few recursive algorithms in schema and validator packages of unnecessary complexity that need to be refactored
